src/viam/sdk/tests/test_base.cpp

```diff
@@ -0,0 +1,136 @@
+#define BOOST_TEST_MODULE test module test_base
+#include <typeinfo>
+#include <unordered_map>
+#include <utility>
+#include <vector>
+
+#include <boost/test/included/unit_test.hpp>
+
+#include <viam/api/common/v1/common.pb.h>
+#include <viam/api/component/base/v1/base.grpc.pb.h>
+#include <viam/api/component/base/v1/base.pb.h>
+
+#include <viam/sdk/common/linear_algebra.hpp>
+#include <viam/sdk/common/proto_type.hpp>
+#include <viam/sdk/components/base/base.hpp>
+#include <viam/sdk/components/base/client.hpp>
+#include <viam/sdk/components/base/server.hpp>
+#include <viam/sdk/tests/mocks/mock_base.hpp>
+#include <viam/sdk/tests/test_utils.hpp>
+
+namespace viam {
+namespace sdktests {
+
+using namespace base;
+
+using namespace viam::sdk;
+
+BOOST_AUTO_TEST_SUITE(test_base)
+
+// This sets up the following architecture
+// -- MockComponent
+//        /\
+//
+//        | (function calls)
+//
+//        \/
+// -- ComponentServer (Real)
+//        /\
+//
+//        | (grpc InProcessChannel)
+//
+//        \/
+// -- ComponentClient (Real)
+//
+// This is as close to a real setup as we can get
+// without starting another process
+//
+// The passed in lambda function has access to the ComponentClient
+//
+template <typename Lambda>
+void server_to_mock_pipeline(Lambda&& func) {
```

Comment:
I know :)

Here, BTW, is what my version of it looks like in the ml model stuff:

```
template <typename F>
void client_server_test(std::shared_ptr<MockMLModelService> mock, F&& f) {
    const auto k_service_name = "mock_mlmodel_service";

    MLModelServiceServer mlmodel_server;
    mlmodel_server.resource_manager()->add(k_service_name, std::move(mock));

    grpc::ServerBuilder builder;
    builder.RegisterService(&mlmodel_server);
    auto server = builder.BuildAndStart();

    MLModelServiceClient client(k_service_name, server->InProcessChannel({}));
    std::forward<F>(f)(client);

    server->Shutdown();
}
```

There are three types in play here: `MockMLModelService`, `MLModelServiceServer`, and `MLModelServiceClient`.

If the mocks had a canonical way of naming the associated client and server types, then this could be templated on the mock type:

```
template <typename Mock, typename F>
void client_server_test(std::shared_ptr<Mock> mock, F&& f) {
    const auto k_service_name = "mocky mock";

    Mock::server_type server;
    server.resource_manager()->add(k_service_name, std::move(mock));

    grpc::ServerBuilder builder;
    builder.RegisterService(&server);
    auto server = builder.BuildAndStart();

    Mock::client_type client(k_service_name, server->InProcessChannel({}));
    std::forward<F>(f)(client);

    server->Shutdown();
}
```

Then you would have:

```
class FooMock {
public:
    using client_type = FooClient;
    using server_type = FooServer;
};
```

That protocol would let you lift this function to a test utility header.

I expect there are further typedefs along these lines that we will want, to make it possible to get easily from a Server type to its Client type and vice versa, and from the Client type to the base type. But I haven't fully formalized how I want that to work. I'm hoping to explore it further in the ml model tests I'm writing.

Reply:
Glad to hear! I'll keep copy/pasting for now but looking forward to simplifying in the future.